api/client/deals.go

```diff
@@ -136,3 +136,44 @@ func (d *Deals) Watch(ctx context.Context, proposals []cid.Cid) (<-chan WatchEve
 	}()
 	return channel, nil
 }
+
+// Retrieve is used to fetch data from filecoin
+func (d *Deals) Retrieve(ctx context.Context, waddr string, cid cid.Cid) (io.Reader, error) {
+	req := &pb.RetrieveRequest{
+		Address: waddr,
+		Cid:     cid.String(),
+	}
+	stream, err := d.client.Retrieve(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	reader, writer := io.Pipe()
+
+	go func() {
+		for {
+			// Check if the context has been canceled, and if so, close the writer an finish
+			_, ok := <-ctx.Done()
+			if !ok {
+				_ = writer.Close()
+				break
+			}
```

Comment:
Trying to think of the best way to respond to the context being canceled. This just checks each time around the loop. Is there a better way?

Reply:
You can do `if ctx.Err() == nil`, if that's the case the context isn't cancelled.

Reply:
Yea? Err gets set if you cancel?

Reply:
Ah I see that in the docs.

Reply:
Yep.
Here's some detailed info:
```
    // If Done is not yet closed, Err returns nil.
    // If Done is closed, Err returns a non-nil error explaining why:
    // Canceled if the context was canceled
    // or DeadlineExceeded if the context's deadline passed.
    // After Err returns a non-nil error, successive calls to Err return the same error.
```

Reply:
Maybe I'm being too picky here, but my thinking about why it sucks to just check for canceled context on each time around the loop is that... if the context is canceled while you're blocked on receiving the stream, you have to wait until the stream is done being received and processed before you check the context state again. I'm working on a way where you'd get notified of the cancelation faster. Not sure how important this distinction is.

Reply:
`stream.Rcv()` shouldn't fail when `ctx` is canceled?
Are you trying to handle that particular case for a reason?

Reply:
I mean, it seems that if it was canceled closing with error would be OK.

Reply:
The use case is the user of the api cancels the context that they passed in to `Retrieve()`. In that case, we should cancel all the work we're doing and return from the function.

Reply:
I see, but I mean in that use-case, wouldn't `reply, err := stream.Recv()` fail with an error and quite with already existing code? Maybe I'm missing something about how `stream` is connected with the API call that returned it.

Reply:
Ah good call. Yes, I think you're right. Because we're passing that user context into `stream, err := d.client.Retrieve(ctx, req)`.

Reply:
Ok so in that case, I don't even need to check if the context is canceled because `stream.Recv()` should just return an error if it is canceled.

Reply:
Yep!